### PR TITLE
pythonPackages.pydiscourse: init at 1.4.0

### PR DESCRIPTION
**Pull request by Dettorer**
## Description of changes

Add the  [`pydiscourse`](https://github.com/pydiscourse/pydiscourse/tree/master) Python module for Python 3.8, 3.9 and 3.10

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Dettorer** [comment]
Result of `nixpkgs-review pr 252145` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python310Packages.pydiscourse</li>
    <li>python310Packages.pydiscourse.dist</li>
    <li>python311Packages.pydiscourse</li>
    <li>python311Packages.pydiscourse.dist</li>
  </ul>
</details>

File: pkgs/development/python-modules/pydiscourse/default.nix [hunk 1]
```diff
@@ -0,0 +1,34 @@
+{
+  buildPythonPackage,
+  fetchFromGitHub ,
+  lib,
+  pythonAtLeast,
+  requests,
+}:
```
Comment (fabaff) [inline pkgs/development/python-modules/pydiscourse/default.nix:7]:
```suggestion
{ lib
, buildPythonPackage
, fetchFromGitHub
, requests
}:
```

`pythonAtLeast` is not used.

File: pkgs/development/python-modules/pydiscourse/default.nix [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+buildPythonPackage rec {
+  pname = "pydiscourse";
+  version = "1.4.0";
+  format = "setuptools";
+
+  # the tox.ini file sets the tests up for Python 3.[7-10] only
+  patches = [ ./0001-Enable-the-tests-for-Python-3.11.patch ];
```
Comment (fabaff) [inline pkgs/development/python-modules/pydiscourse/default.nix:15]: The patch is not needed. It looks like that the tests are not sensitive to the Python release. `unittestCheckHook` will help here.

```suggestion
  nativeCheckInputs = [
    unittestCheckHook
  ];
```

File: pkgs/development/python-modules/pydiscourse/default.nix [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    rev = "refs/tags/v${version}";
+    hash = "sha256-peDkXRcD/ieWYWXqv8hPxTSNRXBHcb/3sj/JJSF2RYg=";
+  };
+
+  propagatedBuildInputs = [ requests ];
+
+  pythonImportsCheck = [ "pydiscourse" ];
+
+  meta = with lib; {
+    homepage = "https://github.com/pydiscourse/pydiscourse";
```
Comment (fabaff) [inline pkgs/development/python-modules/pydiscourse/default.nix:29]:
```suggestion
    homepage = "https://github.com/pydiscourse/pydiscourse";
    changelog = "https://github.com/pydiscourse/pydiscourse/releases/tag/v${version}";
```

File: pkgs/development/python-modules/pydiscourse/default.nix [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    description = "A Python library for working with Discourse.";
```
Comment (fabaff) [inline pkgs/development/python-modules/pydiscourse/default.nix:30]:
```suggestion
    description = "A Python library for working with Discourse";
```

No period at the end please.

File: pkgs/development/python-modules/pydiscourse/default.nix [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  lib,
+  pythonAtLeast,
+  requests,
+}:
+
+buildPythonPackage rec {
+  pname = "pydiscourse";
+  version = "1.4.0";
+  format = "setuptools";
+
```
Comment (fabaff) [inline pkgs/development/python-modules/pydiscourse/default.nix:13]: Supporting EOL releases Python doesn't make much sense for new packages as the chance is high that nobody will test it. Upstream still mention an [addition requirement](https://github.com/pydiscourse/pydiscourse/blob/master/setup.cfg#L21C1-L21C33) for Python < 3.6 but the classifiers only mention Python 3.7.

```suggestion

  disabled = pythonOlder "3.7";
```

**Comment by Dettorer** [comment]
Thanks a lot for the review @fabaff! I truly helped me get some details more clearly in my mind regarding Python packaging for nixpkgs in general.

**Comment by Dettorer** [comment]
Result of `nixpkgs-review pr 252145` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python310Packages.pydiscourse</li>
    <li>python310Packages.pydiscourse.dist</li>
    <li>python311Packages.pydiscourse</li>
    <li>python311Packages.pydiscourse.dist</li>
  </ul>
</details>

**Comment by fabaff** [comment]
Result of `nixpkgs-review pr 252145` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python310Packages.pydiscourse</li>
    <li>python310Packages.pydiscourse.dist</li>
    <li>python311Packages.pydiscourse</li>
    <li>python311Packages.pydiscourse.dist</li>
  </ul>
</details>

**Review by fabaff — APPROVED** [review]
Looks good to me.

**Pull request merged**